Migrating from dockershim documentation (#25787)
kubernetes/website: content/en/docs/tasks/administer-cluster/migrating-from-dockershim/ migrating-telemetry-and-security-agents.md

---
title: Migrating telemetry and security agents from dockershim
content_type: task 
reviewers:
- SergeyKanzhelev
- jrcamp # for SignalFx (Splunk)
weight: 70
---

<!-- overview -->

Docker is being deprecated. From the [Dockershim Deprecation
FAQ](https://kubernetes.io/blog/2020/12/02/dockershim-faq/) you might already
know that most apps do not have a direct dependency on runtime hosting
containers. However, there are still a lot of telemetry and security agents
that has a dependency on docker to collect containers metadata, logs and
metrics. This document aggregates information on how to detect these
dependencies and links on how to migrate these agents to use generic tools or

> alternative runtimes.

Reviewer: To help pages stand the test of time, try to write wording that works both during and after the deprecation. It's not always possible but when it can be done it avoids work to track and reword things after the fact.

I've seen quite a few dated pages on the Kubernetes docs so I try to word assuming that the cleanup probably gets forgotten.

## Telemetry and security agents

> There are a few ways agents may be run on kubernetes cluster. Agents have a direct dependency on Docker when they run on nodes or as daemonsets. Some telemetry and security agents may rely on certain logs or metrics specific for Docker that would not be available any longer. This document does not cover these cases.

Reviewer: nit: “Kubernetes” is a proper noun; we write object kinds like DaemonSet in UpperCamelCase.

### Why do telemetry agents talk to Docker?

Historically, kubernetes was built on top of Docker. Kubernetes is managing networking and scheduling, Docker was placing and operating containers on a node. So you can get scheduling-related metadata like a pod name from Kubernetes and containers state information from Docker. Over time more runtimes were created to manage containers. Also there are projects and kubernetes features that generalize container status information extraction across many runtimes.

> But some agents are still Docker-dependent. They may run commands like [docker
> ps](https://docs.docker.com/engine/reference/commandline/ps/) or [docker

Reviewer: ```suggestion
Some agents are tied specifically to the Docker tool. The agents may run commands like [`docker
 ps`](https://docs.docker.com/engine/reference/commandline/ps/) or [`docker
```
(etc)
?

top](https://docs.docker.com/engine/reference/commandline/top/) to list
containers and processes or [docker
logs](https://docs.docker.com/engine/reference/commandline/logs/) to subscribe
on docker logs. With the deprecating of Docker as a container runtime, these
commands will not work any longer.

> ### Detecting Docker dependency from the agents-daemonsets

Reviewer: This is a task page, so I suggest
```suggestion
### Identify DaemonSets that depend on Docker {#identify-docker-dependency}
```

> If a pod wants to make calls to the `dockerd` running on the node, the pod must mount the docker daemon as a volume.

Reviewer: ```suggestion
If a pod wants to make calls to the `dockerd` running on the node, the pod must either:
- mount the filesystem containing the Docker daemon's privileged socket, as a {{< glossary_tooltip text="volume" term_id="volume" >}}; or
- mount the specific path of the Docker daemon's privileged socket directly, also as a volume.
```
?

> For example, On COS images, docker exposes it’s unix socket at

Reviewer: nits:
```suggestion
For example: on COS images, Docker exposes its Unix domain socket at
```

> `/var/run/docker.sock` This means that the customer’s pod spec will include a

Reviewer: can we assume that there is a customer?

Author: fixed

`hostPath` volume mount of `/var/run/docker.sock`.

> The script to find all pods mounting this volume may look like this:

Reviewer: ```suggestion
Here's a sample shell script to find Pods that have a mount directly
mapping the Docker socket:
```

```bash
kubectl get pods --all-namespaces \
-o=jsonpath='{range .items[*]}{"\n"}{.metadata.namespace}{":\t"}{.metadata.name}{":\t"}{range .spec.volumes[*]}{.hostPath.path}{", "}{end}{end}' \
| sort \
| grep '/var/run/docker.sock'
```

(this script outputs the namespace and name of the pod. You can remove the grep `/var/run/docker.sock` to review other mounts)

> Note, there are alternative ways to call a Docker. For instance, the parent
> directory `/var/run` may be mounted instead of the full path (like in [this
> example](https://gist.github.com/itaysk/7bc3e56d69c4d72a549286d98fd557dd)). So
> the script above will only detect the most common uses.

Reviewer: ```suggestion
{{< note >}}
There are alternative ways for a pod to access Docker on the host. For instance, the parent
directory `/var/run` may be mounted instead of the full path (like in [this
example](https://gist.github.com/itaysk/7bc3e56d69c4d72a549286d98fd557dd)).
The script above only detects the most common uses.
{{< /note >}}
```

Is this acceptable for a task page? If we only explain some of the easier cases, readers might feel short changed.

### Detecting Docker dependency from node deployed agents

In case your cluster nodes are customized and install additional security and
telemetry agents on the node, make sure to check with the vendor of the agent whether it has dependency on Docker.

## Telemetry and security agent vendors

This section is intended to collect information about various telemetry and
security agents that may have dependency on containers runtime.

This section is work in progress.

{{< note >}}

Vendors, please help to fill out this section. See
[CONTRIBUTING](https://github.com/kubernetes/website/blob/master/CONTRIBUTING.md)
guide for instructions. Or create an issue if you need help formatting the
document.

> {{< /note >}}

Reviewer: I would omit this, replacing it with a draft pull request once the other documentation has landed. We don't publish work-in-progress documentation.

Alternatively, hyperlink to a wiki page run by SIG Node, or something similar. Linking to incomplete content is OK.

Author: Would the list of vendors who likely has dependency on Docker be OK to have alongside with this request? Let me draft it...

> ### SignalFx (Splunk)

Reviewer: We prefer to hyperlink to vendor docs (see https://kubernetes.io/docs/contribute/style/content-guide/ for policy). Is that feasible? I think if we let vendors know that this is our policy, vendors ought to be quite willing to add the documentation and to let us know its URL.

If a vendor does not wish to publish documentation, that is fine, we can omit the link.

Author: @jrcamp what do you think?
